### DISCUSSION TITLE
Questions about the difference between pyro-based versus old versions of cell2location

### DISCUSSION THREAD

**Discussion by ciciQ777** [BayraktarLab/cell2location#71] [Q&A] [answered]
Hi @vitkl,

Thanks for creating this wonderful package! I just noticed that there is an updated version of cell2location which was integrated with scvi tools. When I tried the latest tutorial on the example data you provided (Mapping human lymph node cell types to 10X Visium), for the model training part on the scRNAseq reference, it seems that it will take about 30 mins to train when I use GPU. Is this a normal time for this procedure? 

Another question is since I can run it smoothly on the previous version of cell2location, will the cell type proportion estimates different between the newest version and the old version (that uses theano)?

Also is cell2location applicable to higher resolution spatial transcriptomic data which contains spatial locations more than 50k? 

Thank you so much for your time!

**Comment by vitkl** [comment, marked as answer]
Hi @ciciQ777 

1. Training duration of 25-30 min for 250 epochs on the reference dataset of this size is expected.

2. We have some evidence that suggests that the pyro-based version has better numerical accuracy. This can make the precise estimates deviate from the old version, there could also be differences for hard-to-map cell types and low-quality spatial data. However, in general, the results for both the mouse brain and the human lymph node are very similar if not identical.

3. In principle, cell2location is applicable to such higher resolution spatial transcriptomic data. However, a larger data size means that it will be impossible to fit all data into GPU memory. To address this, you need to do training in mini-batches of data (possible with our scvi-tools implementation -`model.train(..., batch_size=2500)`), however, the training time is substantially increased (to days) because cell abundance estimate for each individual location is no longer updated in every training iteration. To address this we can use the approach called amortised inference, in which a neural network is used to approximate cell abundance as a function of gene expression, this leads to improved training efficiency and training duration similar to standard cell2location. The current version provides an experimental implementation of amortised inference. There are not that many datasets with more than 50k locations so we didn't have a chance to test this inference mode comprehensively. Please get in touch if you want to use it and follow updates.

I will move this to discussions since this is not an issue with the package.
